### DISCUSSION TITLE
DefaultFormatter fails to initialize correctly when a frozen application is run as a service on Windows

### DISCUSSION THREAD

**Discussion by johan-ronnkvist** [Kludex/uvicorn#2248] [Potential Issue]
### Context:
I'm starting a uvicorn server as a service on Windows, through `win32serviceutil.ServiceFramework`. The application has been packaged/frozen with PyInstaller.

When starting the application as a service, it crashes with the following stack information 
```
File "uvicorn\main.py", line 522, in run
File "uvicorn\config.py", line 276, in __init__
File "uvicorn\config.py", line 384, in configure_logging
File "logging\config.py", line 912, in dictConfig
File "logging\config.py", line 563, in configure
ValueError: Unable to configure formatter 'default'
``` 

### Additional information:
[This discussion](https://groups.google.com/g/pyinstaller/c/w4EXpztCafE) suggests that the issue comes from `sys.stdout` and `sys.stderr` being set to None in this context.

>Hi Alisha,
> I have seen this problem before and eventually figured out the cause. You are directly or indirectly using either stdout or stderr as an out put for logger but pyinstaller and some or most exe builders produce exes which, when running in Windowed mode, set both of these to None and logger fails like this because None as the exception says doesn’t have an istty or any other methods.
> Some possible solutions include:
> Set the application to be a console application so that it will open a terminal window (it can still use the GUI but log messages will go to the terminal quick but dirty
> - If the application has the frozen attribute or set sys.stdout is None and no log file is specified then create your logger with the file set to os.devnull this will discard all logs
> - Use [logging.NullHandler](https://docs.python.org/3/library/logging.handlers.html#nullhandler) if frozen or sys.stdout is None.
> - Always log to a file but please consider using [logging.RotatingFileHandler](https://docs.python.org/3/library/logging.handlers.html#rotatingfilehandler) as the default handler if frozen or sys.stdout is None.
>Hope that this helps.
>Steve

I have verified that these are `None` for this context, in my own code that executed prior to the above mentioned stack trace.
```
if sys.stderr is None:
    raise AssertionError("sys.stderr is None")
if sys.stdout is None:
    raise AssertionError("sys.stdout is None")
```

### Resolution:
From inspecting the initialization of logging, with the context and information above - I realized that I could bypass the usage of `sys.stdout.isatty()` and `sys.stderr.isatty()` by providing the argument `use_colors` when calling `uvicorn.run()`. This has shown to resolve my crash on startup.

### Suggested change:
I would suggest that no access to `sys.stderr` or `sys.stdout` is allowed before verifying that it's not `None` 

### Related discussions:
[ValueError: Unable to configure formatter 'default'](https://github.com/encode/uvicorn/discussions/2187). Similar error message, but likely a different root cause.

**Comment by TOTom** [comment]
I encounter this problem too ,  when I use the pythonservice.exe running it. how do you resolve it?

**Reply by johan-ronnkvist** [reply]
What part of the text above, that describes how to work around the problem, did you want me to clarify?

**Comment by davidgilbertson** [comment]
I just hit this bug too, just running a server with `pythonw.exe` (not even in a service). Do these words get read by the uvicorn team or should this be an issue? It seems like a legit bug (false assumption that `stdout` has an `isatty()` method) that should be fixed.
